[Streams] Add name and confirm step to create classic stream flyout

damian-polewski opened the pull request:
## Summary

Closes #241644 

This PR adds the step to name and confirm step to the create classic stream flyout.

### How to test
View the component in Storybook:

```bash
yarn storybook classic_stream_flyout
```

| Design | Implementation |
|--------|--------|
| <img width="2880" height="2048" alt="509460451-d7a44513-c959-4a40-800d-7c9b1a1e9b49" src="https://github.com/user-attachments/assets/d44f17c3-4e7f-4341-82df-32acb66efb8b" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 54 16" src="https://github.com/user-attachments/assets/8e136ceb-049e-493a-8981-332ba1b8fa1d" /> |
| <img width="2880" height="2048" alt="509460566-0c38f9f7-d4af-4cdd-850e-9a3142a150a0" src="https://github.com/user-attachments/assets/885fd711-e466-4702-bf41-053b0e2cf869" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 54 28" src="https://github.com/user-attachments/assets/b088419c-7f96-4079-9d1b-ae842bf077af" /> |
| <img width="2880" height="2048" alt="509461037-14f8ef37-ccc3-438b-b6b0-5c1c2bc53a3e" src="https://github.com/user-attachments/assets/41030098-12e5-45b2-9948-21c2857372d8" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 54 36" src="https://github.com/user-attachments/assets/5e8ffcd7-e236-4eee-8ce3-afbbbfbeed08" /> |
| <img width="2880" height="2048" alt="509461130-6de6854a-0f2b-4630-8a2d-1109b09857db" src="https://github.com/user-attachments/assets/9a5e5168-fe1e-40f8-aaec-fdad4687f77a" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 54 45" src="https://github.com/user-attachments/assets/3973e45a-75f4-4054-b45f-21394be94604" /> |
| <img width="2880" height="2048" alt="509461659-5be67d3b-b5f3-42fe-b5a7-743afc960f8f" src="https://github.com/user-attachments/assets/63976e1c-47b0-4e7d-8738-a88b448e52da" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 55 19" src="https://github.com/user-attachments/assets/7c3f941f-573b-48d2-a680-41a601d372f0" /> | 
| <img width="2880" height="2048" alt="509462012-a49a7f21-4d2a-4ff1-8fbd-5b27bb3a202b" src="https://github.com/user-attachments/assets/e55a034b-c102-434e-bd25-4bfc1138fd06" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 55 26" src="https://github.com/user-attachments/assets/1ddb7367-7d5f-4b4b-a11d-539fefeeab90" /> |
| <img width="2880" height="2048" alt="509462190-9ab365ab-b6a0-4148-a8ca-ac53b26b76b9" src="https://github.com/user-attachments/assets/bee4f3e6-a142-4d7b-abb0-891d528f3b3e" /> | <img width="1620" height="916" alt="Screenshot 2025-12-09 at 13 55 34" src="https://github.com/user-attachments/assets/9566a0cc-c249-46f6-8502-63d86a336d9d" /> | 

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

damian-polewski committed:
feat: add name and confirm step

damian-polewski committed:
feat: use prepend/append for static text in the stream name input instead readonly inputs

damian-polewski committed:
feat: move stream name input to separate folder with separate stories

damian-polewski committed:
fix: issue with building stream name

damian-polewski committed:
feat: extract stream name input tests into a separate test file

damian-polewski committed:
fix: folder name typo

damian-polewski committed:
chore: remove template details section

damian-polewski committed:
chore: reorganize tests

damian-polewski committed:
fix: change title size

damian-polewski committed:
fix: only show empty inputs as invalid when validating

sabarasaba commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
```diff
@@ -0,0 +1,143 @@
+  onStreamNameChange: (streamName: string) => void;
+  validationError: ValidationErrorType;
+  conflictingIndexPattern?: string;
+}
```
These types were already defined in `name_and_confirm/name_and_confirm_step.tsx` any chance we can refactor them into a types.ts file and then consume from both places? If we need an extra prop we can always create a new interface and extend the base one

sabarasaba commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx:
```diff
@@ -46,29 +50,99 @@ const flyoutBodyStyles = css({
+    async (name: string): Promise<boolean> => {
+      setIsValidating(true);
+      try {
+        const result = await validateStreamName(name, onValidate);
```
if this fails, we are silently swallowing the error since we dont have a catch. Is this something we should worry about?

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts:
```diff
@@ -0,0 +1,87 @@
+ * Returns the validation result with error type and optional conflicting pattern.
+ */
+export const validateStreamName = async (
+  streamName: string,
```
we are currently passing only `streamName`

But we also support 'higherPriority' validation.

To validate that I assume validator would need to know the priority of the currently selected template to compare against other matching templates?

damian-polewski replied:
Addressed in https://github.com/elastic/kibana/pull/244998/commits/5ba6909ba7f51c2516db3260f52cb6759fc517c8

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx:
```diff
@@ -0,0 +1,260 @@
+      css={getInputGroupStyles()}
+      data-test-subj={dataTestSubj}
+    >
+      {inputGroups.map((group, index) => {
```
do I get it correctly that if an indexPattern has no wildcards like `foo-bar` this will render an empty flex group?

damian-polewski replied:
Now when there is an index pattern with no wildcards like `logs-apache.access`, we will just show a read only input. Addressed in https://github.com/elastic/kibana/pull/244998/commits/5ba6909ba7f51c2516db3260f52cb6759fc517c8

<img width="665" height="168" alt="Screenshot 2025-12-04 at 19 32 44" src="https://github.com/user-attachments/assets/b2761865-c385-4b92-9136-d36f9c6dc729" />

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts:
```diff
@@ -0,0 +1,87 @@
+  if (hasEmptyWildcards(streamName)) {
+    return { errorType: 'empty' };
+  }
+
```
up to you but maybe we could add illegal char validation also from `@kbn/data-view-validation`

To save you time, copy and `pbpaste | git apply` and check for yourself:

```diff
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx
index 66080d656b66..3da447fd9fcf 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx
@@ -90,6 +90,7 @@ export const CreateClassicStreamFlyout = ({
   const [selectedIndexPattern, setSelectedIndexPattern] = useState<string>('');
   const [validationError, setValidationError] = useState<ValidationErrorType>(null);
   const [conflictingIndexPattern, setConflictingIndexPattern] = useState<string | undefined>();
+  const [illegalCharacters, setIllegalCharacters] = useState<string[] | undefined>();
   const [hasAttemptedSubmit, setHasAttemptedSubmit] = useState(false);
   const [isValidating, setIsValidating] = useState(false);
 
@@ -103,6 +104,7 @@ export const CreateClassicStreamFlyout = ({
         const result = await validateStreamName(name, onValidate);
         setValidationError(result.errorType);
         setConflictingIndexPattern(result.conflictingIndexPattern);
+        setIllegalCharacters(result.illegalCharacters);
         return result.errorType === null;
       } finally {
         setIsValidating(false);
@@ -134,6 +136,7 @@ export const CreateClassicStreamFlyout = ({
     setSelectedIndexPattern('');
     setValidationError(null);
     setConflictingIndexPattern(undefined);
+    setIllegalCharacters(undefined);
     setHasAttemptedSubmit(false);
   }, [selectedTemplate]);
 
@@ -141,6 +144,7 @@ export const CreateClassicStreamFlyout = ({
   useEffect(() => {
     setValidationError(null);
     setConflictingIndexPattern(undefined);
+    setIllegalCharacters(undefined);
     setHasAttemptedSubmit(false);
   }, [selectedIndexPattern]);
 
@@ -160,6 +164,7 @@ export const CreateClassicStreamFlyout = ({
       const result = await validateStreamName(streamName, onValidate);
       setValidationError(result.errorType);
       setConflictingIndexPattern(result.conflictingIndexPattern);
+      setIllegalCharacters(result.illegalCharacters);
 
       if (result.errorType === null) {
         onCreate(streamName);
@@ -218,6 +223,7 @@ export const CreateClassicStreamFlyout = ({
             onStreamNameChange={setStreamName}
             validationError={validationError}
             conflictingIndexPattern={conflictingIndexPattern}
+            illegalCharacters={illegalCharacters}
           />
         );
       }
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_and_confirm_step.tsx b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_and_confirm_step.tsx
index 5c50ecda2698..293dec93bbea 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_and_confirm_step.tsx
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_and_confirm_step.tsx
@@ -20,6 +20,7 @@ interface NameAndConfirmStepProps {
   onStreamNameChange: (streamName: string) => void;
   validationError: ValidationErrorType;
   conflictingIndexPattern?: string;
+  illegalCharacters?: string[];
 }
 
 export const NameAndConfirmStep = ({
@@ -29,6 +30,7 @@ export const NameAndConfirmStep = ({
   onStreamNameChange,
   validationError = null,
   conflictingIndexPattern,
+  illegalCharacters,
 }: NameAndConfirmStepProps) => {
   const indexPatterns = template.indexPatterns ?? [];
 
@@ -41,6 +43,7 @@ export const NameAndConfirmStep = ({
         onStreamNameChange={onStreamNameChange}
         validationError={validationError}
         conflictingIndexPattern={conflictingIndexPattern}
+        illegalCharacters={illegalCharacters}
       />
     </EuiFlexGroup>
   );
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx
index 0af164c5cc64..ea06d188a6ed 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx
@@ -15,7 +15,8 @@ import { StreamNameInput } from '../../../stream_name_input';
 
 const getValidationErrorMessage = (
   validationError: ValidationErrorType,
-  conflictingIndexPattern?: string
+  conflictingIndexPattern?: string,
+  illegalCharacters?: string[]
 ) => {
   if (validationError === 'empty') {
     return i18n.translate(
@@ -49,6 +50,20 @@ const getValidationErrorMessage = (
     ];
   }
 
+  if (validationError === 'illegalCharacters' && illegalCharacters) {
+    return [
+      <FormattedMessage
+        key="illegalCharactersError"
+        id="xpack.createClassicStreamFlyout.nameAndConfirmStep.illegalCharactersValidationError"
+        defaultMessage="The stream name contains the invalid {characterListLength, plural, one {character} other {characters}} {characterList}."
+        values={{
+          characterList: <strong>{illegalCharacters.join(' ')}</strong>,
+          characterListLength: illegalCharacters.length,
+        }}
+      />,
+    ];
+  }
+
   return undefined;
 };
 
@@ -59,6 +74,7 @@ interface NameStreamSectionProps {
   onStreamNameChange: (streamName: string) => void;
   validationError: ValidationErrorType;
   conflictingIndexPattern?: string;
+  illegalCharacters?: string[];
 }
 
 export const NameStreamSection = ({
@@ -68,6 +84,7 @@ export const NameStreamSection = ({
   onStreamNameChange,
   validationError,
   conflictingIndexPattern,
+  illegalCharacters,
 }: NameStreamSectionProps) => {
   const { euiTheme } = useEuiTheme();
 
@@ -123,7 +140,11 @@ export const NameStreamSection = ({
         )}
         fullWidth
         isInvalid={validationError !== null}
-        error={getValidationErrorMessage(validationError, conflictingIndexPattern)}
+        error={getValidationErrorMessage(
+          validationError,
+          conflictingIndexPattern,
+          illegalCharacters
+        )}
         helpText={
           <FormattedMessage
             id="xpack.createClassicStreamFlyout.nameAndConfirmStep.nameHelpText"
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts
index bb1491477aa9..02b08a59f0b1 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/utils/utils.ts
@@ -5,6 +5,7 @@
  * 2.0.
  */
 
+import { validateDataView } from '@kbn/data-view-validation';
 import type { TemplateDeserialized } from '@kbn/index-management-plugin/common/types';
 
 export const formatDataRetention = (template: TemplateDeserialized): string | undefined => {
@@ -35,7 +36,12 @@ export const hasEmptyWildcards = (streamName: string): boolean => {
 /**
  * Validation error types for stream name validation
  */
-export type ValidationErrorType = 'empty' | 'duplicate' | 'higherPriority' | null;
+export type ValidationErrorType =
+  | 'empty'
+  | 'duplicate'
+  | 'higherPriority'
+  | 'illegalCharacters'
+  | null;
 
 /**
  * Result from stream name validation
@@ -45,6 +51,8 @@ export interface StreamNameValidationResult {
   errorType: ValidationErrorType;
   /** For higherPriority errors, the conflicting index pattern */
   conflictingIndexPattern?: string;
+  /** For illegalCharacters errors, the list of illegal characters found */
+  illegalCharacters?: string[];
 }
 
 /**
@@ -56,19 +64,35 @@ export type StreamNameValidator = (streamName: string) => Promise<{
 }>;
 
 /**
- * Validates a stream name by checking for empty wildcards and running async validation.
- * Returns the validation result with error type and optional conflicting pattern.
+ * Validates a stream name by checking for empty wildcards, illegal characters, and running async validation.
+ * Returns the validation result with error type and optional details.
  */
 export const validateStreamName = async (
   streamName: string,
   onValidate?: StreamNameValidator
 ): Promise<StreamNameValidationResult> => {
-  // First, check for empty wildcards (local validation)
+  // 1. Check for empty wildcards (local validation)
   if (hasEmptyWildcards(streamName)) {
     return { errorType: 'empty' };
   }
 
-  // Run the external validator if provided
+  // 2. Check for illegal characters (standard Kibana validation)
+  const validationErrors = validateDataView(streamName);
+  if (validationErrors.ILLEGAL_CHARACTERS && validationErrors.ILLEGAL_CHARACTERS.length > 0) {
+    return {
+      errorType: 'illegalCharacters',
+      illegalCharacters: validationErrors.ILLEGAL_CHARACTERS,
+    };
+  }
+
+  if (validationErrors.CONTAINS_SPACES) {
+    return {
+      errorType: 'illegalCharacters',
+      illegalCharacters: [' '],
+    };
+  }
+
+  // 3. Run the external validator if provided
   if (onValidate) {
     const result = await onValidate(streamName);
 
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/tsconfig.json b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/tsconfig.json
index 946c261a9432..37ce32fa522a 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/tsconfig.json
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/tsconfig.json
@@ -11,5 +11,6 @@
     "@kbn/i18n-react",
     "@kbn/storybook",
     "@kbn/index-management-plugin",
+    "@kbn/data-view-validation",
   ]
 }
```

damian-polewski committed:
Merge branch 'main' into feat/add_name_and_confirm_step

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx:
```diff
@@ -0,0 +1,260 @@
+
+        return (
+          <EuiFlexItem
+            key={`wildcard-${group.wildcardIndex}`}
```
wdyt about adding auto focus both here and above for first input item:

> pbpaste | git apply

```diff
diff --git a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx
index aeaf7081a52b..59ac29eb6c4b 100644
--- a/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx
+++ b/x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx
@@ -212,6 +212,7 @@ export const StreamNameInput = ({
     const group = inputGroups[0];
     return (
       <EuiFieldText
+        autoFocus
         value={parts[group.wildcardIndex] ?? ''}
         onChange={(e) => handleWildcardChange(group.wildcardIndex, e.target.value)}
         placeholder="*"
@@ -243,6 +244,7 @@ export const StreamNameInput = ({
             css={getConnectedInputStyles(isFirst, isLast)}
           >
             <EuiFieldText
+              autoFocus={isFirst}
               value={parts[group.wildcardIndex] ?? ''}
               onChange={(e) => handleWildcardChange(group.wildcardIndex, e.target.value)}
               placeholder="*"

```

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368929)
* Commit: ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/368929#019ae90c-1bc5-4f33-8f85-01c08ee06830)

### History
* :broken_heart: [Build #368747](https://buildkite.com/elastic/kibana-pull-request/builds/368747) failed 87e30f033b08e6cc14ab710730c664ece9a0edb5

cc @damian-polewski

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368929","number":368929,"commit":"ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368747","number":368747,"commit":"87e30f033b08e6cc14ab710730c664ece9a0edb5"}],"number":368929}
buildkite-pr-comment-->

sabarasaba commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx:
```diff
@@ -78,6 +152,23 @@ export const CreateClassicStreamFlyout = ({
+    setIsValidating(true);
+
+    try {
+      const result = await validateStreamName(streamName, onValidate);
```
if this fails, we are silently swallowing the error since we dont have a catch. Is this something we should worry about?

sabarasaba reviewed (commented):
Nice work @damian-polewski! Tested the new flyout locally with storybook locally and it all seems to work just fine. I left a few comments, let me know what you think

damian-polewski committed:
fix: cr changes

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369185)
* Commit: 5ba6909ba7f51c2516db3260f52cb6759fc517c8

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/369185#019aeaa5-b88b-4ec4-8e24-d0a55fa99445)

### History
* :broken_heart: [Build #368929](https://buildkite.com/elastic/kibana-pull-request/builds/368929) failed ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c
* :broken_heart: [Build #368747](https://buildkite.com/elastic/kibana-pull-request/builds/368747) failed 87e30f033b08e6cc14ab710730c664ece9a0edb5

cc @damian-polewski

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369185","number":369185,"commit":"5ba6909ba7f51c2516db3260f52cb6759fc517c8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368929","number":368929,"commit":"ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368747","number":368747,"commit":"87e30f033b08e6cc14ab710730c664ece9a0edb5"}],"number":369185}
buildkite-pr-comment-->

damian-polewski committed:
fix: move state to reducer hook

damian-polewski commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx:
```diff
@@ -46,29 +50,99 @@ const flyoutBodyStyles = css({
+  const [validationError, setValidationError] = useState<ValidationErrorType>(null);
+  const [conflictingIndexPattern, setConflictingIndexPattern] = useState<string | undefined>();
+  const [hasAttemptedSubmit, setHasAttemptedSubmit] = useState(false);
+  const [isValidating, setIsValidating] = useState(false);
```
Definitely agree it got messy, good idea - updated in https://github.com/elastic/kibana/pull/244998/commits/025945cf07da1c486274a19ad3a7cfc1cd0cc132

damian-polewski committed:
fix: move name and confirm interface to a separate file

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369468)
* Commit: ad7c71debe03711713abdb62809c6c7b8b2fb9fa

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/369468#019aee59-a049-4074-a59f-7ff19accd579)

### History
* :broken_heart: [Build #369429](https://buildkite.com/elastic/kibana-pull-request/builds/369429) failed 025945cf07da1c486274a19ad3a7cfc1cd0cc132
* :broken_heart: [Build #369185](https://buildkite.com/elastic/kibana-pull-request/builds/369185) failed 5ba6909ba7f51c2516db3260f52cb6759fc517c8
* :broken_heart: [Build #368929](https://buildkite.com/elastic/kibana-pull-request/builds/368929) failed ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c
* :broken_heart: [Build #368747](https://buildkite.com/elastic/kibana-pull-request/builds/368747) failed 87e30f033b08e6cc14ab710730c664ece9a0edb5

cc @damian-polewski

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369468","number":369468,"commit":"ad7c71debe03711713abdb62809c6c7b8b2fb9fa"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369429","number":369429,"commit":"025945cf07da1c486274a19ad3a7cfc1cd0cc132"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369185","number":369185,"commit":"5ba6909ba7f51c2516db3260f52cb6759fc517c8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368929","number":368929,"commit":"ca2ebb2d1ac26e28066dcc043f7ede36011f3a7c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368747","number":368747,"commit":"87e30f033b08e6cc14ab710730c664ece9a0edb5"}],"number":369468}
buildkite-pr-comment-->

damian-polewski committed:
feat: move stream name input utils into a separate file and add unit tests

kapral18 committed:
chore: improve suggestion for https://github.com/elastic/kibana/pull/244998/files

damian-polewski committed:
fix: treat consecutive wildcards as a single wildcard

damian-polewski commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx:
```diff
@@ -0,0 +1,260 @@
+  isLast: boolean;
+}
+
+const parseIndexPattern = (pattern: string): PatternSegment[] => {
```
Probably made a mistake with the logic here treating consecutive wildcards as a separate wildcards e.g. `**-logs-**` wasn't treated as `*-logs-*` and created unnecessary inputs. I fixed this in https://github.com/elastic/kibana/pull/244998/commits/7c4a49acecf56661949c9badc0272f4268a92efd.

kapral18 committed:
chore: remove redundancy from tests

kapral18 committed:
chore: simplify ref usage

damian-polewski committed:
feat: add format validation

kapral18 committed:
chore: reduce ref count and unify validation logic

kapral18 committed:
chore: unifying reducers into a single formik style form state

kapral18 committed:
chore: fix improve validation flow for classic stream flyout
- Refactor `useStreamValidation` hook to introduce a 'Live Validation Mode'
- In 'Live Validation Mode', validation errors are kept visible while typing and debounced validation is triggered
- When a user changes the stream name during an ongoing 'Create' validation, the 'Create' validation is aborted, and the flyout returns to an 'Idle Mode' (no live validation automatically triggered unless there's an existing error)
- Reduce validation delays in stories and tests for faster feedback during development and testing
- Update `formReducer` to correctly set `isSubmitting` to `false` when validation starts, making state more accurate during validation

kapral18 committed:
chore: update validation behavior

- Replaced `getByText` with `findByText` for async validation message
checks to improve test reliability.
- Modified debounced validation logic to ensure it triggers correctly in
"Live Validation Mode" after an initial submission error.
- Updated error visibility during live validation: errors now remain
visible while validation is in progress, instead of clearing immediately
on user input.
- Clarified test descriptions for validation scenarios to better reflect
the introduced "Live Validation Mode."
- Ensured `hasAttemptedSubmit` state is reset correctly when validation
passes in live validation mode, preventing unnecessary subsequent
debounced validations.

kapral18 committed:
chore: fix stories file and remove redundant useEffect

kapral18 committed:
chore: move buildStreamName and countWildcards to utils

- Move utility functions from component to shared utils
- Fix idle mode: abort Create validation returns to idle (not live)
- Remove unused isValid variable

kapral18 committed:
chore: use discriminated union for validation state

Replaces flat validation properties with a type-safe discriminated union
to enforce state machine transitions and make illegal states
unrepresentable.

- Replace
validationMode/isValidating/validationError/conflictingIndexPattern with
single validation object
- Update all action handlers to return validation objects conforming to
union types
- Improve comment clarity and precision
- Reduce test story delay from 5000ms to 2500ms

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx:
```diff
@@ -0,0 +1,178 @@
+  useEffect(() => {
+    const streamName = buildStreamName(indexPattern, parts);
+    onChange(streamName);
+  }, [indexPattern, parts, onChange]);
```
both of these can be dissolved since `useEffect()` should only be used in very specific narrow set of use-cases and should not be used in 98% of cases well outlined here https://react.dev/learn/you-might-not-need-an-effect

There a couple of options: lift state up or use key={indexPattern} from parentComponent to reset identity of component

streamName is a typical derived state that can live at the top of function body and onChange can be called conditionally when `indexPattern` changes inline in the body as they recommend it in the react docs.

In my rewrite suggestion https://github.com/damian-polewski/kibana/pull/2/files I'm actually rolling with lifting state up. But see for yourself which one you like more

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_and_confirm_step.tsx:
```diff
@@ -0,0 +1,35 @@
+      />
+    </EuiFlexGroup>
+  );
+};
```
I didn't change this in my PR suggestion either because it would make the rewrite even more involved than it already is but still for future and general knowledge there is this concept that I really like 
https://julesblom.com/writing/parents-owners-data-flow (I would advice to read all of jules articles, every single one is a golden nugget)

So when you read it I think it will become clear that by introducing these tiny proxy components they only make the prop drill path longer and don't really introduce anything useful.

Read and  let me know if you agree and wdyt about the general philosophy.

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/create_classic_stream_flyout.tsx:
```diff
@@ -46,29 +50,99 @@ const flyoutBodyStyles = css({
+
+  // Debounced validation - only runs after first submit attempt with an error
+  // When validation passes, reset to "submit only" mode
+  useDebounce(
```
I added a long explanation comment in https://github.com/damian-polewski/kibana/pull/2/files#diff-ef9610be27996d164ca5783ccc3662b70176eecc0541de8b25d5c78ec8a15254 outlining drawbacks of each available and considered solution

damian-polewski committed:
Merge branch 'main' into feat/add_name_and_confirm_step

damian-polewski committed:
Merge branch 'feat/add_name_and_confirm_step' into pr_refactor/add_name_and_confirm_step

damian-polewski committed:
fix: pass selected template to onValidate, add validation_utils tests

kapral18 committed:
refactor: update stream name handling in useStreamValidation hook
- Refactor `handleStreamNameChange` to directly update the stream name and trigger validation.
- Remove `lastStreamNameRef` as it's no longer necessary.
- Introduce `setStreamName` as a public method to allow direct updates to the stream name without triggering validation logic in `handleStreamNameChange`.
- Update `updateIndexPattern` to use `setStreamName` instead of `handleStreamNameChange` to avoid redundant validation.
- Adjust `useStreamValidation` to rely on the `formState.streamName` for current stream name, rather than `lastStreamNameRef`.
- Ensure debounced validation explicitly checks against `formState.streamName`.
- Update test cases to include checks that debounced validation does not reintroduce errors after a successful clear.

damian-polewski committed:
feat: add confirm template details section

damian-polewski committed:
Merge pull request #2 from damian-polewski/pr_refactor/add_name_and_confirm_step

chore: improve suggestion for 244998

damian-polewski committed:
fix: add ilm utils unit tests, add abort controller to ilm fetch

damian-polewski committed:
Merge pull request #3 from damian-polewski/feat/add_confirm_template_details_section

feat: add confirm template details section

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

damian-polewski committed:
fix: add aria-label to stream name input

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/confirm_template_details_section.tsx:
```diff
@@ -0,0 +1,242 @@
+        }
+      })
+      .catch((error) => {
+        // Silently fail - we'll just not show phases
```
what if user assumes that policy has no phases, instead of that it's a failure. Maybe we should show a warning toast or error toast https://eui.elastic.co/docs/components/display/toast/#warning ?

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/confirm_template_details_section.tsx:
this component has a dense functionality, this needs tests

damian-polewski replied:
Addressed in https://github.com/elastic/kibana/pull/244998/commits/2c185e116d0c4376c66d22ea3888774a907a5331

kapral18 commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
can we add tests for this one as well

kapral18 reviewed (commented):
Great progress, almost there. Added some final comments.

MichaelMarcialis commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/confirm_template_details_section.tsx:
```diff
@@ -0,0 +1,242 @@
+        description: (
+          <>
+            {componentTemplates.map((ct, idx) => (
+              <div key={idx}>{ct}</div>
```
Currently, each component template name is wrapped in a `div` with no vertical space in between. Would it be possible to add 4px of vertical space between each template name to better match the designs?

damian-polewski replied:
Addressed in https://github.com/elastic/kibana/commit/26bda2f932938ffc44bbdfe86fc7bb695fb63943

MichaelMarcialis reviewed (approved):
Thanks, @damian-polewski! This looks great. I've left some small comments below for your review. Otherwise, once those are addressed, this looks good from my perspective. As I'm out on leave the rest of the week, I'll approve now so I don't hold you up any further.

damian-polewski committed:
fix: design review changes

damian-polewski commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx:
```diff
@@ -0,0 +1,180 @@
+            key={`wildcard-${group.wildcardIndex}`}
+            css={getConnectedInputStyles(isFirst, isLast)}
+          >
+            <EuiFieldText
```
Addressed in https://github.com/elastic/kibana/commit/26bda2f932938ffc44bbdfe86fc7bb695fb63943

damian-polewski committed:
fix: store ilm policy in state and drive phases from it

damian-polewski committed:
fix: extract retention details to separate component

damian-polewski committed:
fix: list items type coercion

damian-polewski commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/confirm_template_details_section.tsx:
```diff
@@ -0,0 +1,242 @@
+          templateDetails as Array<{
+            title: NonNullable<React.ReactNode>;
+            description: NonNullable<React.ReactNode>;
+          }>
```
Good catch! Addressed in https://github.com/elastic/kibana/pull/244998/commits/4c0b36e03add3efff8b9d36f1c9127ec3eb16226

damian-polewski committed:
fix: refactore template details styles

damian-polewski committed:
fix: move index mode labels to commons and use them in components details section

damian-polewski committed:
fix: show error if loading ilm policy fails

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/370933)
* Commit: 59914054a6e266aacd33985b3a5d7b9060f307b0
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-244998)

### Failed CI Steps
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/370933#019b0797-7260-45e4-8dd7-4006bf5deb1b)
* [Jest Tests #17](https://buildkite.com/elastic/kibana-pull-request/builds/370933#019b07b1-8895-4821-aed0-d6ac9fdbef44)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370933#019b0797-7260-45e4-8dd7-4006bf5deb1b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370933/jobs/019b0797-7260-45e4-8dd7-4006bf5deb1b/artifacts/019b07b1-649c-4fc5-9075-f917aaedf829) Jest Tests #<span></span>17 / CreateClassicStreamFlyout name and confirm step displays correct index mode for different templates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/370933#019b07b1-8895-4821-aed0-d6ac9fdbef44) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/370933/jobs/019b07b1-8895-4821-aed0-d6ac9fdbef44/artifacts/019b07c8-7b56-471b-ab09-a501e3d40772) Jest Tests #<span></span>17 / CreateClassicStreamFlyout name and confirm step displays correct index mode for different templates

### History
* :green_heart: [Build #370864](https://buildkite.com/elastic/kibana-pull-request/builds/370864) succeeded 4c0b36e03add3efff8b9d36f1c9127ec3eb16226
* :green_heart: [Build #370519](https://buildkite.com/elastic/kibana-pull-request/builds/370519) succeeded c0fe040bd48a1a5dd5cf74deb53caeb2147050b5
* :green_heart: [Build #369925](https://buildkite.com/elastic/kibana-pull-request/builds/369925) succeeded 85ff0dc0d7f3e14d31d014b23717d3d2398d7425
* :broken_heart: [Build #369636](https://buildkite.com/elastic/kibana-pull-request/builds/369636) failed 74ede91025a739f0f03166520da3354fce097445
* :broken_heart: [Build #369489](https://buildkite.com/elastic/kibana-pull-request/builds/369489) failed 7c4a49acecf56661949c9badc0272f4268a92efd
* :broken_heart: [Build #369483](https://buildkite.com/elastic/kibana-pull-request/builds/369483) failed b578ef36a82cbfdd5f2ab9b76c15338f0096ccab

cc @damian-polewski

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370933","number":370933,"commit":"59914054a6e266aacd33985b3a5d7b9060f307b0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370864","number":370864,"commit":"4c0b36e03add3efff8b9d36f1c9127ec3eb16226"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/370519","number":370519,"commit":"c0fe040bd48a1a5dd5cf74deb53caeb2147050b5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369925","number":369925,"commit":"85ff0dc0d7f3e14d31d014b23717d3d2398d7425"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369636","number":369636,"commit":"74ede91025a739f0f03166520da3354fce097445"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369489","number":369489,"commit":"7c4a49acecf56661949c9badc0272f4268a92efd"}],"number":370933}
buildkite-pr-comment-->

damian-polewski committed:
feat: add confrim details section unit tests

damian-polewski committed:
feat: add name stream section unit tests

damian-polewski committed:
Merge branch 'main' into feat/add_name_and_confirm_step

kapral18 reviewed (approved):
Solid work!

damian-polewski committed:
fix: unnecessary coercion

florent-leborgne commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/confirm_template_details_section.tsx:
```diff
@@ -0,0 +1,301 @@
+            'xpack.createClassicStreamFlyout.nameAndConfirmStep.ilmPolicyErrorDescription',
+            {
+              defaultMessage:
+                'There was an error while loading the ILM policy phases. Please try again later.',
```
```suggestion
                'There was an error while loading the ILM policy phases. Try again later.',
```

florent-leborgne commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
```diff
@@ -0,0 +1,147 @@
+      'xpack.createClassicStreamFlyout.nameAndConfirmStep.emptyValidationError',
+      {
+        defaultMessage:
+          'Please supply a valid text string for all wildcards within the selected index pattern.',
```
```suggestion
          'You must specify a valid text string for all wildcards within the selected index pattern.',
```

yetanothertw replied:
Does one supply a valid string by entering it? If so, using `enter` seems like a more direct and actionable (less ambiguous) alternative.

```suggestion
          'Please enter a valid text string for all wildcards within the selected index pattern.',
```

florent-leborgne commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
```diff
@@ -0,0 +1,147 @@
+      'xpack.createClassicStreamFlyout.nameAndConfirmStep.invalidFormatValidationError',
+      {
+        defaultMessage:
+          'Stream name cannot include \\, /, *, ?, ", <, >, |, comma, #, colon, or spaces. It cannot start with -, _, +, or .ds-. It also cannot be . or ..',
```
Is it possible to format these special characters so they look like "`\`, `/`, `*`, `?`, `"`, `<`, `>`, `|`, `,`, `#`, `:`, or spaces"? I remember having a similar comment on another Streams-related PR and I think this was done somehow

florent-leborgne commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
```diff
@@ -0,0 +1,147 @@
+    return i18n.translate(
+      'xpack.createClassicStreamFlyout.nameAndConfirmStep.duplicateValidationError',
+      {
+        defaultMessage: 'This stream name already exists. Please try a different name.',
```
```suggestion
        defaultMessage: 'This stream name already exists. Try a different name.',
```

florent-leborgne reviewed (approved):
Copy LGTM - a few styling nits and one formatting request though I don't think I'll need to take another look at this after :D 

Thanks!

yetanothertw reviewed (approved):
LGTM, thank you! 

I've left a couple of tiny suggestions for clarity, but will leave it up to you to make the call.

damian-polewski committed:
fix: copy changes

damian-polewski commented on x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/create_classic_stream_flyout/steps/name_and_confirm/name_stream_section.tsx:
```diff
@@ -0,0 +1,147 @@
+      <FormattedMessage
+        key="higherPriorityError"
+        id="xpack.createClassicStreamFlyout.nameAndConfirmStep.higherPriorityValidationError"
+        defaultMessage="This stream name matches a higher priority index template with index pattern {pattern}. Please alter your stream name so that it no longer conflicts, or consider changing your selected index template."
```
Adressed in https://github.com/elastic/kibana/pull/244998/commits/60703b7752a9ff0fbda309c6e5ec1be1a6df7eee

damian-polewski committed:
fix: adjust unit tests after copy change

damian-polewski committed:
fix: copy changes 2

damian-polewski committed:
fix: index mode label import path

damian-polewski committed:
Revert "fix: move index mode labels to commons and use them in components details section"

This reverts commit 7652565b300ba7347d6cd4ab4c152ca43299ef11.

damian-polewski committed:
fix: add index mode labels to utils
